### PR TITLE
Make AbstractBlock extend AbstractRefCounted

### PR DESCRIPTION
**Pull request by original-brownbear**
The functionality here is just duplicated and wasn't thread-safe, which seems broken given the way this is used.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java [hunk 1]
```diff
@@ -100,43 +100,6 @@ public boolean isReleased() {
...
-        if (isReleased()) {
-            return false;
-        }
-        references++;
-        return true;
-    }
-
-    @Override
-    public final boolean decRef() {
-        if (isReleased()) {
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:122]: If we use `AbstractRefCounted`, we'll have to update some use of vectors, which sometimes can get released without us decrementing the refcount all the way, but still releasing their containing blocks. Currently, blocks generated from `X-VectorBlock` override `isReleased` to account for this.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Uff this is a mess, we have a ref counted `AbstractBlock` wrapping `BooleanVector` which is releaseable. If we close the wrapped vector, then the abstract block stops working and just quietly returns `false` on `tryIncRef` which is not in line with how ref counting works normally. This makes it almost impossible to make the block implementation work with #102551, we need to find some way of fixing this stuff up and bringing it in line with how we normally do ref-counting. This is blocking #102030.

**Review by alex-spies — COMMENTED** [review]
@original-brownbear , thanks for your refactor!

However, we decided not to use `AbstractRefCounted`, mostly because we do not need thread safety. (A block is either being used in an operator, run by a single thread, or being sent to another thread - never both.)

I can see the added simplicity of just extending `AbstractRefCounted`, though.

**Comment by alex-spies** [comment]
Pushed changes which should make the tests pass + avoid states where vector blocks are released even if their refcount is >0.

**Comment by alex-spies** [comment]
Discussed with Chris and Armin: having a thread safe refcount (and making sure that released == refcount 0) should add simplicity and no significant overhead.

I'd wait for @nik9000 and @dnhatn to see this before merging, though.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/IntBlockHash.java [hunk 2]
```diff
@@ -59,20 +59,20 @@ public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
                     addInput.add(0, groupIds);
                 }
             } else {
-                try (IntVector groupIds = add(intVector)) {
-                    addInput.add(0, groupIds);
+                try (IntBlock groupIds = add(intVector)) {
+                    addInput.add(0, groupIds.asVector());
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/IntBlockHash.java:63]: asVector() here should return a non-null value here. However, this change is unsafe, as asVector() is allowed to return null. I think we should inline this or use `finally` to close the Block.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/LongBlockHash.java [hunk 3]
```diff
@@ -62,20 +62,20 @@ public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
                     addInput.add(0, groupIds);
                 }
             } else {
-                try (IntVector groupIds = add(longVector)) {
-                    addInput.add(0, groupIds);
+                try (IntBlock groupIds = add(longVector)) {
+                    addInput.add(0, groupIds.asVector());
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/LongBlockHash.java:66]: Same comment here for asVector

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/DoubleBlockHash.java [hunk 4]
```diff
@@ -62,20 +62,20 @@ public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
                     addInput.add(0, groupIds);
                 }
             } else {
-                try (IntVector groupIds = add(doubleVector)) {
-                    addInput.add(0, groupIds);
+                try (IntBlock groupIds = add(doubleVector)) {
+                    addInput.add(0, groupIds.asVector());
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/DoubleBlockHash.java:66]: Same comment for `asVector`

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRefBlockHash.java [hunk 5]
```diff
@@ -68,20 +68,20 @@ public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
                     addInput.add(0, groupIds);
                 }
             } else {
-                try (IntVector groupIds = add(bytesVector)) {
-                    addInput.add(0, groupIds);
+                try (IntBlock groupIds = add(bytesVector)) {
+                    addInput.add(0, groupIds.asVector());
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRefBlockHash.java:72]: `asVector`

**Review by dnhatn — COMMENTED** [review]
I am +1 to switch to thread-safe reference counting, considering its minimal overhead. However, I would like Nik to decide this.

Regarding the Vector issue, I think we should also implement refCounted for Vector, and the `asBlock` method should create a new Block that has its own reference to the Vector. Callers should close both Vector and Block appropriately.

**Comment by alex-spies** [comment]
> I am +1 to switch to thread-safe reference counting, considering its minimal overhead. However, I would like Nik to decide this.
> 
> Regarding the Vector issue, I think we should also implement refCounted for Vector, and the `asBlock` method should create a new Block that has its own reference to the Vector. Callers should close both Vector and Block appropriately.

++ to having `asBlock` create a new Vector; I think Vectors really shouldn't have a reference to their enclosing blocks as this makes reasoning about ownership quite complicated.

IMO, if we can get away with it, I'd like vectors to not even be refcounted, only Blocks. I think this should work, and for optimizations like not creating a deep copy of vectors for filtering, we may be able to go through the owning blocks instead.

Before we proceed with this one, I think we should be clear on how to handle vectors; may be let's leave this open until we tackle avoiding deep copies when filtering.

**Comment by original-brownbear** [comment]
closing this after discussing with @alex-spies , this introduces performance overhead that is hard to quantify and isn't needed for anything right now.

**Pull request closed without merging**